WIP: Add Scraper/ScraperRun classes

tmtmtmtm opened the pull request:
First stab at factoring out a Scraper class and a ScraperRun class.

The IndexToMemebers class is probably overkill for now — we can find the
higher level abstractions like that after we nail down the basics — but
it's probably useful to include in this intial WIP which is more for
feedback than expected to actually merge yet.

tmtmtmtm committed:
WIP: Add Scraper/ScraperRun classes

First stab at factoring out a Scraper class and a ScraperRun class.

The IndexToMemebers class is probably overkill for now — we can find the
higher level abstractions like that after we nail down the basics — but
it's probably useful to include in this intial WIP which is more for
feedback than expected to actually merge yet.

chrismytton commented:
Part of https://github.com/everypolitician/everypolitician/issues/572.

chrismytton commented:
The `ScraperRun` class is part of https://github.com/everypolitician/everypolitician/issues/574.

chrismytton committed:
Raise error after saving to runs table

This gives us the full stack trace and ensures that the program actually
exits after an error.

chrismytton committed:
Fix bug in index_fields_from

This was trying to call `#keys` on an array, which was raising an error.
For now just use the first item from the data array.

chrismytton committed:
Remove start_date from default_index_fields

Not all records seem to have a start date, so removing this from the
list of default index fields for now.

chrismytton committed:
Make index_fields_from method private

This doesn't need to be a public method as it's only used internally.

chrismytton committed:
Replace IndexToMembers class with RomanianParliament class

Rather than trying to find the correct generic abstraction this forces
the scrapers to have a class that encapsulates navigating a website to
get the correct content.

In the future we can potentially wrap these up in `IndexToMembers` type
classes, but for now it seems easier to just do scraper specific ones
until we find the correct abstraction.

chrismytton committed:
Merge branch 'master' into scraper-class

chrismytton committed:
Better name for scraper class

tmtmtmtm commented:
Thinking about this a little more, I fear that we got a key underlying concept back to front here. Here we say that a `Scraper` has a `ScraperRun`. But I think we're being led astray by two subtly different uses of the world `scraper`. The `Scraper` class here wraps an interaction with `Scraped`, but that's not the same use of a `scraper` in the morph sense, which might (and often does) require *lots* of calls to actually fetch, parse, and extract data from multiple pages. 

I now think this should all be the other way around: a `ScraperRun` should be the primary interface, rather than essentially a private class.
